File: paleta/pages/palette/palette_page.py
from gi.repository import Adw, Gio, Gtk

from .palette_row import PaletteRow
from paleta.model import Model, Palette
from paleta.dialog import AddNewPaletteDialog


@Gtk.Template(resource_path='/io/github/nate_xyz/Paleta/palette_page.ui')
class PalettePage(Adw.Bin):
    __gtype_name__ = 'PalettePage'

    list_box = Gtk.Template.Child(name="list_box")
    status = Gtk.Template.Child(name="status")
    add_palette_button = Gtk.Template.Child(name="add_palette_button")

    def __init__(self) -> None:
        super().__init__()
        self.list_store = Gio.ListStore(item_type=Palette)
        self.list_box.bind_model(self.list_store, self.listbox_factory)
        self.edit_mode = False
        self.window = None
        self.database = None
        self.add_palette_button.connect('clicked', lambda _button: self.show_new_palette_dialog())
        
    def saturate(self, window, database, model: Model):
        self.window = window 
        self.edit_button = window.edit_palette_button
        self.database = database
        self.model = model
        self.model.connect('populated', self.update_view)

    def update_view(self, model=None):
        self.list_store.remove_all()
        palettes = self.model.get_palettes().items()
        if len(palettes) == 0:
           self.status.show()
           self.list_box.hide()
           self.edit_button.set_css_classes(['flat'])
           self.edit_button.hide()
           self.edit_mode = False
        else:
            self.status.hide()
            self.list_box.show()
        for _, palette in palettes:
            self.list_store.append(palette)
        self.update_edit_view()

    def update_edit_view(self):
        for pc in self.list_box:
            pc.edit_mode = self.edit_mode
            pc.update_edit_view()

    def listbox_factory(self, palette):
        return PaletteRow(palette, self.window, self.database, self.model)

    def toggle_edit_mode(self):
        if len(self.list_store) > 0:
            self.set_edit_mode(not self.edit_mode)
            self.window.go_to_palette_page()
        else:
            self.window.add_error_toast(_("Cannot toggle edit mode, no palettes added."))

    def set_edit_mode(self, mode):
        self.edit_mode = mode
        if self.edit_mode:
            self.edit_button.set_css_classes(['opaque', 'edit-action-button'])
        else:
            self.edit_button.set_css_classes(['flat'])
        self.update_edit_view()

    def show_new_palette_dialog(self):
        def after_dialog():
            self.set_edit_mode(False)
            self.edit_button.show()
            self.window.go_to_palette_page()

        dialog = AddNewPaletteDialog(self.window, self.database, self.model)
        dialog.connect('response', lambda dialog, response: after_dialog())
        dialog.show()

File: src/pages/palette/palette_page.rs
use adw::{prelude::*, subclass::prelude::*};
use gtk::{glib, glib::clone, CompositeTemplate, gio};

use std::cell::Cell;

use crate::model::palette::Palette;
use crate::dialog::add_new_palette_dialog::AddNewPaletteDialog;
use crate::toasts::add_error_toast;
use crate::i18n::i18n;
use crate::util::{model, edit_button_set_visible, edit_button_mode};

use super::palette_row::PaletteRow;

mod imp {
    use super::*;

    #[derive(Debug, Default, CompositeTemplate)]
    #[template(resource = "/io/github/nate_xyz/Paleta/palette_page.ui")]
    pub struct PalettePagePriv {
        #[template_child(id = "list_box")]
        pub list_box: TemplateChild<gtk::ListBox>,

        #[template_child(id = "status")]
        pub status: TemplateChild<adw::StatusPage>,

        #[template_child(id = "add_palette_button")]
        pub add_palette_button: TemplateChild<gtk::Button>,

        pub list_store: gio::ListStore,
        pub edit_mode: Cell<bool>,
    }

    #[glib::object_subclass]
    impl ObjectSubclass for PalettePagePriv {
        const NAME: &'static str = "PalettePage";
        type Type = super::PalettePage;
        type ParentType = adw::Bin;

        fn class_init(klass: &mut Self::Class) {
            Self::bind_template(klass);
        }

        fn instance_init(obj: &glib::subclass::InitializingObject<Self>) {
            obj.init_template();
        }

        fn new() -> Self {
            Self {
                list_box: TemplateChild::default(),
                status: TemplateChild::default(),
                add_palette_button: TemplateChild::default(),
                list_store: gio::ListStore::new(Palette::static_type()),
                edit_mode: Cell::new(false),
            }
        }

    }

    impl ObjectImpl for PalettePagePriv {
        fn constructed(&self) {
            self.parent_constructed();
            self.obj().initialize();
        }
    }

    impl WidgetImpl for PalettePagePriv {}
    impl BinImpl for PalettePagePriv {}
    impl PalettePagePriv {}
}

glib::wrapper! {
    pub struct PalettePage(ObjectSubclass<imp::PalettePagePriv>)
    @extends gtk::Widget, adw::Bin;
}


impl PalettePage {
    pub fn new() -> PalettePage {
        let color_panel: PalettePage = glib::Object::builder::<PalettePage>().build();
        color_panel
    }

    fn initialize(&self) {
        let imp = self.imp();

        imp.list_box.bind_model(Some(&imp.list_store), 
        clone!(@strong self as this => @default-panic, move |obj| {
            let palette = obj.clone().downcast::<Palette>().expect("Palette is of wrong type");       
            return PaletteRow::new(palette).upcast::<gtk::Widget>();
            })
        );

        model().connect_local(
            "populated",
            false,
            clone!(@weak self as this => @default-return None, move |_args| {
                this.update_view();
                None
            }),
        );

        imp.add_palette_button.connect_clicked(
            clone!(@strong self as this => @default-panic, move |_button| {
                this.show_new_palette_dialog();
            }),
        );
    }

    fn update_view(&self) {
        let imp = self.imp();
        imp.list_store.remove_all();
        let palettes = model().palettes();
        if palettes.is_empty() {
            imp.status.show();
            imp.list_box.hide();
            edit_button_set_visible(false);
            imp.edit_mode.set(false);
        } else {
            imp.status.hide();
            imp.list_box.show();
            for (_i, palette) in palettes.iter() {
                imp.list_store.append(palette.as_ref());
            }
        }
        self.update_edit_view()
    }

    pub fn toggle_edit_mode(&self) {
        let imp = self.imp();
        if imp.list_store.n_items() > 0 {
            self.set_edit_mode(!imp.edit_mode.get());
        } else {
            add_error_toast(i18n("Cannot toggle edit mode, no palettes added."))
        }
    }

    fn set_edit_mode(&self, mode: bool) {
        let imp = self.imp();
        imp.edit_mode.set(mode);
        edit_button_mode(mode);
        self.update_edit_view();
    }

    fn update_edit_view(&self) {
        let imp = self.imp();
        for row in imp.list_box.observe_children().snapshot() {
            let palette_row = row.downcast::<PaletteRow>().expect("PaletteRow is of wrong type");      
            palette_row.set_edit_mode(imp.edit_mode.get());
        }
    }

    fn show_new_palette_dialog(&self) {
        let dialog = AddNewPaletteDialog::new();
        dialog.connect_response(            None,
            clone!(@strong self as this => move |_dialog, response| {
                if response == "add" {
                    this.set_edit_mode(false);
                    edit_button_set_visible(true);
                }
            }),
        );
        dialog.show()
    }
}
    